### PR TITLE
fix: Handle deletion of unsaved copied file in NotebookPanel

### PR DESCRIPTION
**Pull request by georgecwan**
* Previously, duplicating a notebook file and then trying to delete it without saving it first will cause an error to occur
* Added a `fileExists` function to `FileUtils` to handle this check
* Added a `getUniqueCopyFileName` function to `FileUtils`
* Behaviour of copying a file from `NotebookPanel` changed from opening an unsaved copy of the file to duplicating the file in `FileStorage` first
* Related to #1359 

#### Testing Instructions:
* Open a file in the notebook panel
* Copy the opened file by using the 'Copy File' option in the overflow menu
* Delete the file by using the overflow menu (No error should occur)
* Repeat the same steps but copy the file by right clicking the file in the file explorer
* Delete the copied file again, this time it should also disappear from the file explorer

### CODE REVIEW

**Review by mofojed — CHANGES_REQUESTED** [review]
There's still a case that can result in lost data:
1) Have a file named `foo.py`
2) In File Explorer, right-click the file and select "Copy File". A new file `foo-copy.py` is created.
3) Open foo-copy.py, make some edits, then save and close it.
4) Open the original `foo.py` notebook
5) From the notebook menu, choose "Copy File". It will open a new tab with the name `foo-copy.py`, effectively masking the file from the same name already on disk.
6) From the new tab with the name `foo-copy.py`, select `Delete File`. It will delete the `foo-copy.py` file from disk.

I think the fix for that is in `handleCopy` to loop when getting the copy file name until you know that file doesn't exist.

**Comment by georgecwan** [comment]
That wouldn't fix the problem since the act of copying the file doesn't save it (this was the behaviour before my change) so trying to copy the same file twice without saving either of them first will still have the same problem. Is there a way to get all of files that `NotebookPanel` has open?

**Comment by mofojed** [comment]
Right that's a good point; or you could copy it from the notebook panel, then copy it from the file explorer after.
Yes, there is an `openFileMap` defined in `ConsolePlugin` that keeps track of all the files that are open. You could use redux to set that for the current dashboard (similar to `setDashboardConsoleSettings`) and then use that when copying files...

File: packages/dashboard-core-plugins/src/panels/FileExplorerPanel.tsx
Comment (mofojed) [inline packages/dashboard-core-plugins/src/panels/FileExplorerPanel.tsx:170]:
```suggestion
    // eslint-disable-next-line no-await-in-loop
```
Doesn't seem to be necessary to remove the strict-boolean-expressions warning.

File: packages/dashboard-core-plugins/src/panels/NotebookPanel.tsx
Comment (mofojed) [inline packages/dashboard-core-plugins/src/panels/NotebookPanel.tsx:695]: This whole block could be factored into a method in `FileUtils`, e.g.
```
async function getUniqueCopyFileName(fileStorage, originalName) { ... }
```
We do the exact same thing in FileExplorerPanel

File: packages/dashboard-core-plugins/src/panels/NotebookPanel.tsx
Comment (mofojed) [inline packages/dashboard-core-plugins/src/panels/NotebookPanel.tsx:708]: We should pass `true` for the `shouldFocus` parameter here, so the newly copied panel gets focus (as it did before).

File: packages/dashboard-core-plugins/src/panels/NotebookPanel.tsx
Comment (mofojed) [inline packages/dashboard-core-plugins/src/panels/NotebookPanel.tsx:697]: Ideally we'd be catching an error if one occurred here, and displaying a toast for it. However, we don't really have toasts wired up at all at the moment, and it's possible we just wire up a toast service for catching uncaught exceptions anyway so do not worry about it for now.

**Review by mofojed — COMMENTED** [review]
Make sure we get @dsmmcken 's approval before merging, as this does change behaviour (since the file is created right on copy, instead of requiring an explicit save after). I think it's better overall (too many weird cases the other way).

File: packages/dashboard-core-plugins/src/panels/FileExplorerPanel.tsx
Comment (georgecwan) [reply packages/dashboard-core-plugins/src/panels/FileExplorerPanel.tsx:170]: Without disabling the strict-boolean-expressions check, I get this warning:
image

Should I ignore it then?

Comment (mofojed) [reply packages/dashboard-core-plugins/src/panels/FileExplorerPanel.tsx]: I think your TypeScript server be trippin', if you restart it (Ctrl+P, then type `> Restart TypeScript Server`), it'll fix it. If it's still causing issues just leave it.

**Review by mofojed — APPROVED** [review]

**Review by dsmmcken — APPROVED** [review]

**Pull request merged**
